DIPs/DIP1xxx-JS.md

```diff
@@ -0,0 +1,110 @@
+| Review Count:   | 0 |
+| Author:         | Jack Stouffer |
+| Implementation: | N/A |
+| Status:         | Draft |
+
+## Abstract
+
+In order to incrementally improve D or it's standard library, it's often necessary to
+mark features or functions for future removal. This document proposes a standardized
+process for language maintainers to remove public features. This process would be 
+used across DMD, Druntime, and Phobos.
+
+## Rationale
+
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends up being handled slightly differently depending on
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, so as to minimize breakage and
+to provide clear fixes for user code.
+
+## Description
+
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Users must be given at least 10 non-patch releases before the deprecated features
+are removed. More releases can be given if the removed code is commonly used.
+There are two cases where the deprecation period is allowed to be shorter:
+
+1. The code or feature is notably dangerous or unsafe, and users need to remove
+it from their code as soon as possible.
+2. The existance of the current code precludes its own fix or the fix of an equally
+important issue.
+
+Shorting the deprecation preiod should be done with caution to avoid giving D
+an image of instability.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
+are tested on the Project Tester that are broken should also have their
+maintainers notified.
+
+Both at the time of deprecation and removal, a changelog entry must be made. This
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
+catch all planned deprecations.
+
+### Public Functions, Types, and Modules
+
+All removals (or changes that make the code `private`/`package`) of public functions,
+types, and modules must be accompanied with a deprecation period.
+
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period. A pointer to more information should also be added. E.g.
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
```

Comment:
Ouh so you were talking about this. As a user, this is a bit annoying.
OK, so I have to search for the changelog or function signature and hope I can find it. This takes at least three hops and is unnecessarily challenging.
I love the error messages from ShellCheck, e.g.

https://github.com/koalaman/shellcheck/wiki/SC2206

Can't we emit a similar deprecation message code? Phobos, dmd and druntime could use different leading numbers to avoid sync issues, but if numbers are too problematic because of merge conflicts (bumping the number), short keywords like DEPHEXSTRING or DEPENFORCEEX could work too?

Reply:
>https://github.com/koalaman/shellcheck/wiki/SC2206

Maybe I'm missing something but I don't see anything about error messages on that page.

>short keywords like DEPHEXSTRING or DEPENFORCEEX could work too?

That's basically just an error code, which really isn't helpful either. At the end of the day, to get details on the deprecation/migration path you still would have to go to google and type that in. In the case described in the DIP, at least there's a specific place the user is told to go to.

Reply:
Normally, if you get a deprecation message, then you just look at the documentation for that symbol. I see no reason to complicate this with something like error codes (something that Walter has already expressed a dislike for BTW).

The deprecation message tells you that the symbol has been deprecated, and if the explanation is short enough, it also tells you what to use instead. If you want to know more, you look at the documentation for that symbol. For those who read the changelog, a changelog entry can give a heads up, but ultimately, it's the documentation on the symbol that's telling you what's going on. Let's please not complicate this.

Reply:
For that to work better, it would be convenient to provide persistent link to documentation entry for the specific version used.

Reply:
> For that to work better, it would be convenient to provide persistent link to documentation entry for the specific version used.

Do you mean that the changelog entries should use links like e.g. this one?

https://docarchives.dlang.io/v2.079.0/phobos/std_functional.html#binaryReverseArgs

Reply:
No, I mean actual deprecation messages printed by compiler. Less steps to take from message to relevant instructions -> more developers who actually read instructions as opposed to guessing -> less frustration

Reply:
What @mihails-strasuns-sociomantic requets requires (1) dlang.org keep an archive (lasting at least as long as the oldest deprecations) of all documentation for each release, and (2) maintain a consistent URL scheme that doesn't change over time.  Both are good things, IMO, and we should definitely move in that direction.  Then we can put actual URLs in the deprecation messages without the chaos that will ensue if the contents of the URL changes over time or if the URL itself becomes invalid over time.  (Keep in mind that the URL must work even for previous releases of the compiler, so things like fixing a typo in the URL, etc., need to be done with care, e.g., with a webserver redirect in the right places.)